### PR TITLE
Improve/fix version,status reporting for service-ca clusteroperator

### PR DESCRIPTION
**Pull request by sallyom**
Switch to using library-go/operator v1helpers.UpdateStatus and other improvements in status reporting

Summary of status reporting (3 deployments == configmap-cabundle-injector, apiservice-cabundle-injector, and serving-cert-signer) :
```
Available=True | VersionUpdate=True | Progressing=False
* All replicas of each of the 3 deployments is running
* No instances running from previous deployments

Available=True | VersionUpdate=True | Progressing=True
* At least 1 replica of each of the 3 deployments is running
* No instances running from previous deployments
* Not all replicas of each of the 3 deployments exist, still deploying

Available=True | VersionUpdate=False | Progressing=True
* At least 1 replica of each of the 3 deployments is running
* Instances found running from previous deployments

Available=False 
* At least 1 replica of any of the 3 deployments is not running

Progressing=True
* IsNotFound error any 1 of deployments
* Found DeletionTimestamp for any deployment
* A deployment does not have an available replica
* A deployment does not have it’s configured available replicas

Failing=True
* Error returned from syncLoop
* Error in fetching deployments (but not IsNotFound error)
```

There is a ~10 second gap in version reporting of service-ca CO during upgrades.
 [this](https://github.com/openshift/library-go/pull/319) fixes the gap in version reporting
Bumping deps here: https://github.com/openshift/service-ca-operator/pull/40

### CODE REVIEW

File: pkg/operator/status.go [hunk 1]
```diff
@@ -35,24 +34,13 @@ func (c *serviceCAOperator) setFailingStatus(operatorConfig *operatorv1.ServiceC
...
+		Reason:  reason,
+		Message: message,
 	})
-
-	v1helpers.SetOperatorCondition(&operatorConfig.Status.Conditions,
-		operatorv1.OperatorCondition{
-			Type:   operatorv1.OperatorStatusTypeAvailable,
-			Status: operatorv1.ConditionFalse,
-		})
 }
```
Comment (sallyom) [inline pkg/operator/status.go]: @stlaz I think for setProgressing, we should only set Progressing==True, leave Available as/is.  wdyt? I wish there was a doc to explain what the correct behavior should be.  So, Progressing&Available may be true, however, only if Available=True &&Progressing=False do we report a new version.  That's what I'm thinking.

**Comment by sallyom** [comment]
@mrogers950 @enj please take a look.  I've tested a bunch locally.

File: pkg/operator/status.go [hunk 2]
```diff
@@ -98,55 +94,87 @@ func isDeploymentStatusComplete(deploy *appsv1.Deployment) bool {
...
 		return true, nil
 	}
-	// report Available is deployments and replicas exist, but don't report version
-	// if there are replica instances remaining from previous deployment
+	// Further step down in readiness:
+	// Here, ready to report Available=True
+	// because all deployments exist and at least 1 replicas exists per deployment.
+	// Don't report version here because
+	// there are replica instances remaining from previous deployment
+	// Progressing will be true here
```
Comment (stlaz) [inline pkg/operator/status.go]: `all deployments have at least 1 replica, but some are of previous versions - don't report new version`?

File: pkg/operator/operator.go [hunk 3]
```diff
@@ -136,3 +128,10 @@ func getGeneration(client appsclientv1.AppsV1Interface, ns, name string) int64 {
 	}
 	return deployment.Generation
 }
+
+func (c serviceCAOperator) updateStatus(operatorConfigCopy *operatorv1.ServiceCA) {
```
Comment (mrogers950) [inline pkg/operator/operator.go]: nit: that the argument is a Copy does not seem relevant to the function so I would suggest just operatorConfig

File: pkg/operator/status.go [hunk 4]
```diff
@@ -98,55 +94,87 @@ func isDeploymentStatusComplete(deploy *appsv1.Deployment) bool {
...
 	}
-	// set Available if replica instances are created.
-	// report version if all deployments are available and updated
-	if ready == len(deployments) {
-		glog.Infof("All deployments managed by service-ca-operator are available and updated")
-		c.setAvailableStatus(operatorConfigCopy)
+	// Available, Updated, and Ready to report version:
+	// Here, ready to report version and set Available=True and
+	// set Progressing=False because all deployments, replicas exist
+	// and all instances are updated, no previous deployment instances exist
```
[the remark this answers is not in the document]
Comment (mrogers950) [reply pkg/operator/status.go]: The returning of true to report a version or not kind of confused me as well. I think it would be cleaner for this function to have a structure where you use defer for the status/version update.

File: pkg/operator/operator.go (hunk 3 above)
Comment (sallyom) [reply pkg/operator/operator.go]: done

File: pkg/operator/status.go [hunk 5]
```diff
@@ -98,55 +94,87 @@ func isDeploymentStatusComplete(deploy *appsv1.Deployment) bool {
...
 
 func (c *serviceCAOperator) syncStatus(operatorConfigCopy *operatorv1.ServiceCA, deployments []string) (bool, error) {
-	ready := 0
+	version_ready := 0
 	existingDeploymentsAndReplicas := 0
+	deployment_complete := 0
+	statusMsg := ""
 	for _, dep := range deployments {
+		reason := "ManagedDeploymentsNotReady"
 		existing, err := c.appsv1Client.Deployments(operatorclient.TargetNamespace).Get(dep, metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/status.go]: thinking about this.. working it out

File: pkg/operator/operator.go [hunk 6]
```diff
@@ -15,44 +13,41 @@ import (
...
 )
 
 type serviceCAOperator struct {
-	operatorConfigClient   operatorv1client.ServiceCAsGetter
-	operatorConfigInformer operatorv1informers.ServiceCAInformer
-
-	appsv1Client appsclientv1.AppsV1Interface
-	corev1Client coreclientv1.CoreV1Interface
-	rbacv1Client rbacclientv1.RbacV1Interface
+	operatorClient *operatorclient.OperatorClient
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/operator.go]: done

File: pkg/operator/operator.go [hunk 7]
```diff
@@ -15,44 +11,39 @@ import (
...
+	appsv1Client   appsclientv1.AppsV1Interface
+	corev1Client   coreclientv1.CoreV1Interface
+	rbacv1Client   rbacclientv1.RbacV1Interface
 	versionGetter status.VersionGetter
 	eventRecorder events.Recorder
 }
 
 func NewServiceCAOperator(
-	operatorConfigInformer operatorv1informers.ServiceCAInformer,
+	operatorClient *operatorclient.OperatorClient,
```
Comment (sallyom) [inline pkg/operator/operator.go]: done

File: pkg/operator/status.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/status.go]: I've updated,it's more clear now.  I'm setting the version in the syncStatus.  I'm also passing a list of deployments from the targetNamespace to the syncStatus func.  Not sure if it's better, I'm testing w/ it now.

File: pkg/operator/status.go [hunk 8]
```diff
@@ -98,55 +94,87 @@ func isDeploymentStatusComplete(deploy *appsv1.Deployment) bool {
...
+		reason := "ManagedDeploymentsCompleteAndUpdated"
+		c.setAvailableTrue(operatorConfigCopy, reason)
+		c.setProgressingFalse(operatorConfigCopy, reason, "All service-ca-operator deployments updated")
+		return true, nil
+	}
+	// Step down in readiness:
+	// Here, ready to report a version and set Available=True
+	// because all deployments exist and at least 1 replica exists per deployment,
+	// and there are no instances from previous deployments.
+	// Progressing will be true because deployments aren't complete.
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/status.go]: :+1:

File: pkg/operator/status.go (hunk 2 above)
Comment (sallyom) [reply pkg/operator/status.go]: :+1:

File: pkg/operator/operator.go [hunk 9]
```diff
@@ -84,11 +79,12 @@ func NewServiceCAOperator(
 		operator.WithInformer(namespacedKubeInformers.Core().V1().Secrets(), secretEvents),
 		operator.WithInformer(namespacedKubeInformers.Apps().V1().Deployments(), deploymentEvents),
 		operator.WithInformer(namespacedKubeInformers.Core().V1().Namespaces(), namespaceEvents),
+		operator.WithInformer(operatorClient.Informers.Operator().V1().ServiceCAs(), configEvents),
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/operator.go]: done

File: pkg/operator/operator.go [hunk 10]
```diff
@@ -15,44 +13,41 @@ import (
...
+	appsv1Client   appsclientv1.AppsV1Interface
+	corev1Client   coreclientv1.CoreV1Interface
+	rbacv1Client   rbacclientv1.RbacV1Interface
 
 	versionGetter status.VersionGetter
 	eventRecorder events.Recorder
 }
 
 func NewServiceCAOperator(
 	operatorConfigInformer operatorv1informers.ServiceCAInformer,
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/operator.go]: yes, removed : )

File: pkg/operator/status.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/status.go]: the comments were more for me, to make sure I knew what I was doing, i shortened them

[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/status.go]: :)

File: pkg/operator/status.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/status.go]: I'm listing the existing deps in the targetNS, then passing the list to syncStatus, checking against the expectedDeps. testing that now

File: pkg/operator/operator.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/operator.go]: done

File: pkg/operator/status.go [hunk 11]
```diff
@@ -97,57 +87,84 @@ func isDeploymentStatusComplete(deploy *appsv1.Deployment) bool {
...
-		if err != nil {
-			if apierrors.IsNotFound(err) {
-				statusMsg := fmt.Sprintf("Deployment %s does not exist", dep)
-				c.setProgressingStatus(operatorConfigCopy, statusMsg)
-				return false, nil
+func (c *serviceCAOperator) syncStatus(operatorConfigCopy *operatorv1.ServiceCA, existingDeps *appsv1.DeploymentList, targetDeps []string) {
+	versionReady := true
+	deploymentsComplete := true
+	statusMsg := ""
+	for _,  target := range targetDeps {
```
Comment (stlaz) [inline pkg/operator/status.go]: I'm not quite sure this does what I thought it was going to do, the original might have been a bit better. How about this:

Have 4 strings - `notFound`, `noReplicas`, `notYetComplete` or similar, they correspond to the original conditions (if you're bringing any new here, feel free to add, I did not check fully).

In each of the conditions, if it applies for the deployment, do `<thisConditionStringName> += fmt.Sprintf("\n%s", deploymentName)` and instead of the original `return false, nil`, just do `continue`

After the loop `for _, dep := range deployments {` finishes (keep `dep` here, btw, using shortcuts in loops is fine IMHO), check the string vars you had and do (for example):
```go
progressingMsg := ""
if len(notFound) != 0 {
    progressingMsg += fmt.Sprintf("The following deployments were not yet created:\n%s", notFound)
}
```
and eventually set the `Progressing` and `Available` statuses accordingly, with the progressing status having the long message, available can just be more brief.

Please do reach out to me if any of this is not clear :wink:

File: pkg/operator/status.go [hunk 12]
```diff
@@ -98,55 +94,87 @@ func isDeploymentStatusComplete(deploy *appsv1.Deployment) bool {
 }
 
 func (c *serviceCAOperator) syncStatus(operatorConfigCopy *operatorv1.ServiceCA, deployments []string) (bool, error) {
-	ready := 0
+	version_ready := 0
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/status.go]: got it, done (have to double check my logic, doing that now)

File: pkg/operator/status.go (hunk 11 above)
Comment (sallyom) [reply pkg/operator/status.go]: yes, I am in the process of rechecking the logic, etc.. still wip.  it's clear, thanks : ) also I agree the syncStatus needs rework

File: pkg/operator/status.go [hunk 13]
```diff
@@ -97,57 +87,84 @@ func isDeploymentStatusComplete(deploy *appsv1.Deployment) bool {
...
-	ready := 0
-	existingDeploymentsAndReplicas := 0
-	for _, dep := range deployments {
-		existing, err := c.appsv1Client.Deployments(operatorclient.TargetNamespace).Get(dep, metav1.GetOptions{})
-		if err != nil {
-			if apierrors.IsNotFound(err) {
-				statusMsg := fmt.Sprintf("Deployment %s does not exist", dep)
-				c.setProgressingStatus(operatorConfigCopy, statusMsg)
-				return false, nil
+func (c *serviceCAOperator) syncStatus(operatorConfigCopy *operatorv1.ServiceCA, existingDeps *appsv1.DeploymentList, targetDeps []string) {
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/status.go]: :+1:

File: pkg/operator/operator.go [hunk 14]
```diff
@@ -15,44 +13,40 @@ import (
 	"github.com/openshift/library-go/pkg/operator/status"
 
 	operatorv1 "github.com/openshift/api/operator/v1"
-	operatorv1client "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1"
-	operatorv1informers "github.com/openshift/client-go/operator/informers/externalversions/operator/v1"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/operator.go]: done

File: pkg/operator/status.go (hunk 11 above)
Comment (sallyom) [reply pkg/operator/status.go]: @stlaz , I've cleaned  up syncStatus (had some extra booleans there) : )  now looking at your suggestion, maybe I'll revert back to a cleaner version of what I had b4. 
> Have 4 strings - `notFound`, `noReplicas`, `notYetComplete` or similar, they correspond to the original conditions (if you're bringing any new here, feel free to add, I did not check fully).
> 
> In each of the conditions, if it applies for the deployment, do `<thisConditionStringName> += fmt.Sprintf("\n%s", deploymentName)` and instead of the original `return false, nil`, just do `continue`

Comment (sallyom) [reply pkg/operator/status.go]: @stlaz @mrogers950 I think syncStatus() makes sense the way it is atm.  We are following just 3 deployments, they are all deployed w/ the same operator image, so unless somehow the deployment yaml was modified (b4 install? since it's managed via CVO) the notFound will not happen (unless something is horribly wrong cluster-wide).  For this component once Progressing = true during an upgrade, a status for 1 of the deployments will suffice as overall CO status msg (since they all use same image, managed via CVO), and if progressing=true because a deployment was manually deleted, the correct status will be reported for that deployment. 

I've gathered `oc get serviceca -o yaml` during both an upgrade and when I manually delete the deployments in openshift-service-ca.  The status/reasons/messages all are as expected.

Comment (stlaz) [reply pkg/operator/status.go]: https://github.com/openshift/service-ca-operator/pull/38 makes it possible for the deployments to get redeployed, I suspect `notFound` would still be possible.

Would deployments each be always in the same state, even during upgrade/start? Could replicas of one deployment be slower of the other? I guess they could, but correct me if I'm wrong, I am not familiar with the corner cases of service-ca operator :slightly_smiling_face: 

In both the first case and the second one, though, some replicas may not be found, some may be late and having the information about each the situation for both IMHO makes sense.

Also, to simplify the code a bit, consider moving the `targetDeploymentNames` slice into a set (https://github.com/kubernetes/apimachinery/blob/master/pkg/util/sets/string.go#L30) and use `x.Has(dep)` instead.

Comment (sallyom) [reply pkg/operator/status.go]: > Also, to simplify the code a bit, consider moving the `targetDeploymentNames` slice into a set (https://github.com/kubernetes/apimachinery/blob/master/pkg/util/sets/string.go#L30) and use `x.Has(dep)` instead.
thanks :)

Comment (sallyom) [reply pkg/operator/status.go]: _much better_

File: pkg/operator/operator.go [hunk 15]
```diff
@@ -104,28 +100,24 @@ func (c serviceCAOperator) Sync(obj metav1.Object) error {
...
 		err := syncControllers(c, operatorConfigCopy)
 		if err != nil {
-			c.setFailingStatus(operatorConfigCopy, "OperatorSyncLoopError", err.Error())
-		} else {
-			setOperatorVersion, err = c.syncStatus(operatorConfigCopy, deploymentNames)
-			if err != nil {
-				return err
-			}
+			c.setFailingTrue(operatorConfigCopy, "OperatorSyncLoopError", err.Error())
+			c.updateStatus(operatorConfigCopy)
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/operator.go]: done

File: pkg/operator/operator.go [hunk 16]
```diff
@@ -104,28 +100,24 @@ func (c serviceCAOperator) Sync(obj metav1.Object) error {
...
 			}
 		}
 	}
-	// update status to be available, progressing or failing
-	if !equality.Semantic.DeepEqual(operatorConfig, operatorConfigCopy) {
-		if _, err := c.operatorConfigClient.ServiceCAs().UpdateStatus(operatorConfigCopy); err != nil {
-			return err
-		}
-	}
+	c.updateStatus(operatorConfigCopy)
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator/operator.go]: got it
